### ISSUE TITLE
LDAP sources: searchattributes='' option prevent radius to start

### ISSUE THREAD

**Issue by nqb** [inverse-inc/packetfence#3919]
After doing an upgrade to 8.3 and running `/usr/local/pf/addons/upgrade/to-8.3-authentication-searchattributes.pl` script, users got following in `authentication.conf` for LDAP sources:
```
[..]
searchattributes=
```
When they try to generate RADIUS config, they got following error:
```console
# pfcmd service radiusd generateconfig
service|command
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
```
which prevent RADIUS to start.

Caused by https://github.com/inverse-inc/packetfence/blame/devel/lib/pf/services/manager/radiusd_child.pm#L575

Several ways to fix this issue:
- `/usr/local/pf/addons/upgrade/to-8.3-authentication-searchattributes.pl` could copy/paste value of `usernameattribute` into `searchattributes`
- add conditionals or defaults in `generate_radiusd_ldap` method

**Comment by nqb** [comment]
We must generate new stable packages with fix.

**Comment by julsemaan** [comment]
The warning alone doesn't prevent RADIUS from starting, can you provide the output of radius.log and packetfence.log when its starting.

Technically, this case should have been handled after discussions with @fdurand so its surprising we're hitting this

**Comment by nqb** [comment]
With same `authentication.conf`, issue only present on Debian 8.

On CentOS, with `searchattributes=`, I got:
```console
# pfcmd service radiusd generateconfig
service|command
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
radiusd-acct|config generated
radiusd-auth|config generated

**Comment by julsemaan** [comment]
the above is more inline with what I would expect.

Perhaps the new freeradius package for Debian hasn't been put into the repo and that is causing an issue

**Comment by nqb** [comment]
As discussed with @fdurand, issue is related to a missing `pfcmd configreload hard` before running `pfcmd service radiusd generateconfig` during an upgrade to 8.3.

In UPGRADE.asciidoc, we must add this step to run this command before restarting pf services.

**Comment by qehs244** [comment]
nothing like finding you've got the exact same issue after an upgrade to 8.3 on debian 8 and the issue is only 9 hours old (grin)

**Comment by qehs244** [comment]
root@packetfence:/usr/local/pf/bin# ./pfcmd service radiusd generateconfig
service|command
Use of uninitialized value $searchattributes in concatenation (.) or string at /usr/local/pf/lib/pf/services/manager/radiusd_child.pm line 556.
root@packetfence:/usr/local/pf/bin# ./pfcmd service radiusd restart
Service                                                 Status    PID
packetfence-radiusd-auth.service                  	stopped   0
packetfence-radiusd-acct.service                  	stopped   0
packetfence-radiusd-cli.service                   	stopped   0
Checking configuration sanity...
Job for packetfence-radiusd-auth.service failed. See 'systemctl status packetfence-radiusd-auth.service' and 'journalctl -xn' for details.
packetfence-radiusd-auth.service                  	stopped   0
Job for packetfence-radiusd-acct.service failed. See 'systemctl status packetfence-radiusd-acct.service' and 'journalctl -xn' for details.
packetfence-radiusd-acct.service                  	stopped   0
Job for packetfence-radiusd-cli.service failed. See 'systemctl status packetfence-radiusd-cli.service' and 'journalctl -xn' for details.
packetfence-radiusd-cli.service                   	stopped   0

 Loaded: loaded (/lib/systemd/system/packetfence-radiusd-auth.service; enabled)
   Active: failed (Result: start-limit) since Fri 2019-01-11 00:48:01 GMT; 1min 29s ago
     Docs: man:radiusd(8)
           man:radiusd.conf(5)
           http://wiki.freeradius.org/
           http://networkradius.com/doc/
  Process: 2402 ExecStartPre=/usr/bin/perl -I/usr/local/pf/lib -Mpf::services::manager::radiusd -e pf::services::manager::radiusd->new()->generateConfig() (code=exited, status=255)

Jan 11 00:48:01 packetfence systemd[1]: Failed to start PacketFence FreeRADIUS authentication multi-protocol authentication server.
Jan 11 00:48:01 packetfence systemd[1]: Unit packetfence-radiusd-auth.service entered failed state.
Jan 11 00:48:01 packetfence systemd[1]: packetfence-radiusd-auth.service holdoff time over, scheduling restart.
Jan 11 00:48:01 packetfence systemd[1]: Stopping PacketFence FreeRADIUS authentication multi-protocol authentication server...
Jan 11 00:48:01 packetfence systemd[1]: Starting PacketFence FreeRADIUS authentication multi-protocol authentication server...
Jan 11 00:48:01 packetfence systemd[1]: packetfence-radiusd-auth.service start request repeated too quickly, refusing to start.
Jan 11 00:48:01 packetfence systemd[1]: Failed to start PacketFence FreeRADIUS authentication multi-protocol authentication server.
Jan 11 00:48:01 packetfence systemd[1]: Unit packetfence-radiusd-auth.service entered failed state.

**Comment by qehs244** [comment]
**As a quick fix if you add UserPrincipleName and mail (at least in my case) into the field "search attributes" in the AD authentication source, all radius services start up and seem to function as intended**

**Comment by nqb** [comment]
Hello @qehs244,

The fix added in  b54866e didn't work for you ?

**Comment by qehs244** [comment]
As a quick fix if you add **sAMAccountName** into the field **"search attributes"** in the AD authentication source, all radius services start up and seem to function as intended

**Comment by qehs244** [comment]
> Hello @qehs244,
> 
> The fix added in [b54866e](https://github.com/inverse-inc/packetfence/commit/b54866e787850dd7942f97986f5b1ff2a8d0cf05) didn't work for you ?

no radius still didn't start until I put  **sAMAccountName** in the search attributes field then everything fired up

**Comment by nqb** [comment]
Fixed by bdac1ea7
